src/greenlight/runner.clj

```diff
@@ -77,6 +108,54 @@
     true))
 
 
+(defn- run-group
+  [^ExecutorService exec-pool system tests]
+  {:post [(instance? Reader (first %))
+          (instance? Future (second %))]}
+  (let [reader (PipedReader.)
+        writer (PipedWriter. reader)
+        bindings (get-thread-bindings)
+        group-f (reify Callable
+                  (call [_]
+                    (with-bindings bindings
+                      (binding [*out* writer, *err* writer]
+                        (try
+                          (mapv (partial test/run-test! system) tests)
+                          (finally
+                            (.close writer)))))))]
+    [reader (.submit exec-pool group-f)]))
+
+
+(defn- print-reader
+  [^Reader reader]
+  (with-open [br (BufferedReader. reader)]
+    (loop []
+      (when-let [line (.readLine br)]
+        (println line)
+        (recur)))))
+
+
+(defn- execute-parallel
+  "Run a collection of tests, using an executor pool with `n-threads`"
+  [system tests n-threads]
+  (let [exec-pool (Executors/newFixedThreadPool n-threads)]
+    (try
+      (->> tests
+           (group-by #(or (::test/group %) (gensym)))
+           (map val)
+           (map (partial run-group exec-pool system))
+           (doall)
+           (map
+             (fn print-and-deref
+               [[reader f]]
+               (print-reader reader)
+               (.get ^Future f)))
+           (doall)
+           (mapcat identity))
+      (finally
+        (.shutdownNow exec-pool)))))
```

Comment:
I don't think we're exposed to that since `step/advance!` is wrapped in a try/catch for catching test functions that happen to throw an Exception, to record an `:error` state.